packages/block-editor/src/components/inserter/index.js

```diff
@@ -279,24 +285,35 @@ export default compose( [
 								directInsertBlock &&
 								directInsertBlock?.name === lastInnerBlock.name
 							) {
-								return { ...lastInnerBlock.attributes };
+								adjacentAttributes = lastInnerBlock.attributes;
 							}
 						}
-						return {};
+					} else {
+						// Otherwise, attempt to get attributes from the
+						// previous block relative to the current clientId.
+						const currentBlock = getBlock( clientId );
+						const previousBlock = getBlock(
+							getPreviousBlockClientId( clientId )
+						);
+
+						if ( currentBlock?.name === previousBlock?.name ) {
+							adjacentAttributes =
+								previousBlock?.attributes || {};
+						}
 					}
 
-					// Attempt to get attributes from the previous block
-					// relative to the current clientId.
-					const currentBlock = getBlock( clientId );
-					const previousBlock = getBlock(
-						getPreviousBlockClientId( clientId )
-					);
-
-					if ( currentBlock?.name === previousBlock?.name ) {
-						return { ...( previousBlock?.attributes || {} ) };
+					// Copy over only those attributes flagged to be copied.
+					for ( const attribute in attributesToCopy ) {
+						if (
+							attributesToCopy[ attribute ] &&
```

Comment:
Can we remove this and just check:

`if ( adjacentAttributes.hasOwnProperty( attribute ) )`?

Reply:
Thanks for re-testing @stacimc, another good question! I wasn't too sure about this one — the `attributesToCopy` is currently an object where each attribute is set to a boolean (`true` in all the cases in the Buttons block). So, if someone were to theoretically set one of them to `false` it might be odd if it still copied the attribute.

I initially had it an object so that look up was easy when looping over the adjacent block's attributes, so an object look up was preferable to an array. Now that it's switched the other way (we're looping over the `attributesToCopy`) the object structure doesn't seem to be _quite_ as necessary. I kinda like the parallel, though, with having both `attributes` and `attributesToCopy` being objects.

So, a couple of follow-up questions! 😀

Is it better for `attributesToCopy` to be an object or an array? And if we keep it as an object, should it be possible for someone to set one of the values to `false` to switch off copying of that attribute (there's no real use case for this, but I _think_ it determines whether or not we should remove this line 🤔)

Apologies if I'm overthinking this!

Reply:
Oh of course, `attributesToCopy` is an object now 🤦‍♀️  You're right, this is a bit tricky! Provided `attributesToCopy` remains an object, I'd leave it as you have it. I don't think anyone _ought_ to explicitly set a value to `false` within that object, but I think you're right to safeguard against it.

To me it does feel slightly less clear when reading -- maybe just a naming thing, but I wouldn't expect a parameter named `attributesToCopy` to contain data for attributes that _shouldn't_ be copied. I may be the one overthinking it here, though 😄

Reply:
Making `attributesToCopy` an array would make it less verbose to add those attributes, and semantically makes more sense because we wouldn't ever expect them to be set to `false` (there's no point adding an attribute to copy if we don't want it copied, and there's no way of toggling those attributes after they're set). 

I agree it's not a blocking issue for this PR though; we can always address it in a follow up.

Reply:
Thanks @tellthemachines, that's a great point. I've updated it to be an array in https://github.com/WordPress/gutenberg/pull/37905/commits/401358fc268b6cc74033650a3b203adb032b52b8 (and updated the `typedef` so that it's an array of strings), and that looks much clearer to me. Cheers!

I'm out for the next few hours, but will 🤞 merge in this PR tonight once the tests pass, unless there are any objections 🙂